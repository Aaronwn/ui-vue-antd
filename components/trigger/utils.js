function isPointsEq (a1, a2) {
  return a1[0] === a2[0] && a1[1] === a2[1]
}

export function getAlignFromPlacement (builtinPlacements, placementStr, align) {
  const baseAlign = builtinPlacements[placementStr] || {}
  return {
    ...baseAlign,
    ...align,
  }
}

export function getPopupClassNameFromAlign (builtinPlacements, prefixCls, align) {
  const points = align.points
  for (const placement in builtinPlacements) {
    if (builtinPlacements.hasOwnProperty(placement)) {
      if (isPointsEq(builtinPlacements[placement].points, points)) {
        return `${prefixCls}-placement-${placement}`
      }
    }
  }
  return ''
}
export function noop () {
}
