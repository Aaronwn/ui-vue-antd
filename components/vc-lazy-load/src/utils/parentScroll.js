const style = (element, prop) => {
  let styleVal = ''
  if (typeof getComputedStyle !== 'undefined') {
    styleVal = window.getComputedStyle(element, null).getPropertyValue(prop)
  } else {
    styleVal = element.style[prop]
  }
  return styleVal
}

const overflow = (element) =>
  style(element, 'overflow') + style(element, 'overflow-y') + style(element, 'overflow-x')

const scrollParent = (element) => {
  if (!(element instanceof window.HTMLElement)) {
    return window
  }

  let parent = element

  while (parent) {
    if (parent === document.body || parent === document.documentElement) {
      break
    }

    if (!parent.parentNode) {
      break
    }
    if (/(scroll|auto)/.test(overflow(parent))) {
      return parent
    }

    parent = parent.parentNode
  }

  return window
}

export default scrollParent
