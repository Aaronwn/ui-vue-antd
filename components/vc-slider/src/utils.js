import keyCode from '../../_util/KeyCode'

export function isEventFromHandle (e, handles) {
  console.log(Object.keys(handles)
    .some(key => e.target === handles[key].$el))
  return Object.keys(handles)
    .some(key => e.target === handles[key].$el)
}

export function isValueOutOfRange (value, { min, max }) {
  return value < min || value > max
}

export function isNotTouchEvent (e) {
  return e.touches.length > 1 ||
    (e.type.toLowerCase() === 'touchend' && e.touches.length > 0)
}

export function getClosestPoint (val, { marks, step, min }) {
  const points = Object.keys(marks).map(parseFloat)
  if (step !== null) {
    const closestStep =
            Math.round((val - min) / step) * step + min
    points.push(closestStep)
  }
  const diffs = points.map(point => Math.abs(val - point))
  return points[diffs.indexOf(Math.min(...diffs))]
}

export function getPrecision (step) {
  const stepString = step.toString()
  let precision = 0
  if (stepString.indexOf('.') >= 0) {
    precision = stepString.length - stepString.indexOf('.') - 1
  }
  return precision
}

export function getMousePosition (vertical, e) {
  return vertical ? e.clientY : e.pageX
}

export function getTouchPosition (vertical, e) {
  return vertical ? e.touches[0].clientY : e.touches[0].pageX
}

export function getHandleCenterPosition (vertical, handle) {
  const coords = handle.getBoundingClientRect()
  return vertical
    ? coords.top + (coords.height * 0.5)
    : coords.left + (coords.width * 0.5)
}

export function ensureValueInRange (val, { max, min }) {
  if (val <= min) {
    return min
  }
  if (val >= max) {
    return max
  }
  return val
}

export function ensureValuePrecision (val, props) {
  const { step } = props
  const closestPoint = getClosestPoint(val, props)
  return step === null ? closestPoint
    : parseFloat(closestPoint.toFixed(getPrecision(step)))
}

export function pauseEvent (e) {
  e.stopPropagation()
  e.preventDefault()
}

export function getKeyboardValueMutator (e) {
  switch (e.keyCode) {
    case keyCode.UP:
    case keyCode.RIGHT:
      return (value, props) => value + props.step

    case keyCode.DOWN:
    case keyCode.LEFT:
      return (value, props) => value - props.step

    case keyCode.END: return (value, props) => props.max
    case keyCode.HOME: return (value, props) => props.min
    case keyCode.PAGE_UP: return (value, props) => value + props.step * 2
    case keyCode.PAGE_DOWN: return (value, props) => value - props.step * 2

    default: return undefined
  }
}

export function getComponentProps (obj, prop) {
  if (obj[prop]) {
    return obj
  } else if (obj.$children.length) {
    const len = obj.$children.length
    for (let i = 0; i < len; i++) {
      if (obj.$children[i][prop]) {
        return obj.$children[i]
      } else if (obj.$children[i].$children.length) {
        return getComponentProps(obj.$children[i], prop)
      }
    }
  }
}
