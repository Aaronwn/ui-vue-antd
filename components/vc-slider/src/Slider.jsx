import PropTypes from '../../_util/vue-types'
import warning from '../../_util/warning'
import BaseMixin from '../../_util/BaseMixin'
import { hasProp } from '../../_util/props-util'
import Track from './common/Track'
import createSlider from './common/createSlider'
import * as utils from './utils'

const Slider = {
  name: 'Slider',
  mixins: [BaseMixin],
  props: {
    defaultValue: PropTypes.number,
    value: PropTypes.number,
    disabled: PropTypes.bool,
    autoFocus: PropTypes.bool,
    tabIndex: PropTypes.number,
  },
  data () {
    const defaultValue = this.defaultValue !== undefined
      ? this.defaultValue : this.min
    const value = this.value !== undefined
      ? this.value : defaultValue

    if (process.env.NODE_ENV !== 'production') {
      warning(
        !hasProp(this, 'minimumTrackStyle'),
        'minimumTrackStyle will be deprecate, please use trackStyle instead.'
      )
      warning(
        !hasProp(this, 'maximumTrackStyle'),
        'maximumTrackStyle will be deprecate, please use railStyle instead.'
      )
    }
    return {
      sValue: this.trimAlignValue(value),
      dragging: false,
    }
  },
  mounted () {
    this.$nextTick(() => {
      const { autoFocus, disabled } = this
      if (autoFocus && !disabled) {
        this.focus()
      }
    })
  },
  watch: {
    value: {
      handler (val) {
        const { min, max } = this
        this.setChangeValue(val, min, max)
      },
      deep: true,
    },
    min (val) {
      const { sValue, max } = this
      this.setChangeValue(sValue, val, max)
    },
    max (val) {
      const { sValue, min } = this
      this.setChangeValue(sValue, min, val)
    },
  },
  methods: {
    setChangeValue (value, min, max) {
      const minAmaxProps = {
        min,
        max,
      }
      const newValue = value !== undefined
        ? value : this.sValue
      const nextValue = this.trimAlignValue(newValue, minAmaxProps)
      if (nextValue === this.sValue) return

      this.setState({ sValue: nextValue })
      if (utils.isValueOutOfRange(newValue, minAmaxProps)) {
        this.$emit('change', nextValue)
      }
    },
    onChange (state) {
      const isNotControlled = !hasProp(this, 'value')
      if (isNotControlled) {
        this.setState(state)
      }

      const changedValue = state.sValue
      this.$emit('change', changedValue)
    },
    onStart (position) {
      this.setState({ dragging: true })
      const { sValue } = this
      this.$emit('beforeChange', sValue)

      const value = this.calcValueByPos(position)

      this.startValue = value
      this.startPosition = position
      if (value === sValue) return

      this.prevMovedHandleIndex = 0

      this.onChange({ sValue: value })
    },
    onEnd () {
      this.setState({ dragging: false })
      this.removeDocumentEvents()
      this.$emit('afterChange', this.sValue)
    },
    onMove (e, position) {
      utils.pauseEvent(e)
      const { sValue } = this
      const value = this.calcValueByPos(position)
      if (value === sValue) return

      this.onChange({ sValue: value })
    },
    onKeyboard (e) {
      const valueMutator = utils.getKeyboardValueMutator(e)

      if (valueMutator) {
        utils.pauseEvent(e)
        const { sValue } = this
        const mutatedValue = valueMutator(sValue, this.$props)
        const value = this.trimAlignValue(mutatedValue)
        if (value === sValue) return

        this.onChange({ sValue: value })
      }
    },
    getLowerBound () {
      return this.min
    },
    getUpperBound () {
      return this.sValue
    },
    trimAlignValue (v, nextProps = {}) {
      const mergedProps = { ...this.$props, ...nextProps }
      const val = utils.ensureValueInRange(v, mergedProps)
      return utils.ensureValuePrecision(val, mergedProps)
    },
    getTrack ({ prefixCls, vertical, included, offset, minimumTrackStyle, _trackStyle }) {
      return (
        <Track
          class={`${prefixCls}-track`}
          vertical={vertical}
          included={included}
          offset={0}
          length={offset}
          style={{
            ...minimumTrackStyle,
            ..._trackStyle,
          }}
        />
      )
    },
    renderSlider () {
      const {
        prefixCls,
        vertical,
        included,
        disabled,
        minimumTrackStyle,
        trackStyle,
        handleStyle,
        tabIndex,
        min,
        max,
        handle: handleGenerator,
      } = this
      const { sValue, dragging } = this
      const offset = this.calcOffset(sValue)
      const handle = handleGenerator(this.$createElement, {
        prefixCls,
        vertical,
        offset,
        value: sValue,
        dragging,
        disabled,
        min,
        max,
        index: 0,
        tabIndex,
        style: handleStyle[0] || handleStyle,
        ref: 'handleRefs_0',
        handleFocus: this.onFocus,
        handleBlur: this.onBlur,
      })

      const _trackStyle = trackStyle[0] || trackStyle
      return {
        tracks: this.getTrack({ prefixCls, vertical, included, offset, minimumTrackStyle, _trackStyle }),
        handles: handle,
      }
    },
  },
}

export default createSlider(Slider)
