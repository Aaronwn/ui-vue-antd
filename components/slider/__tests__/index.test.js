import { mount } from '@vue/test-utils'
import { asyncExpect } from '@/tests/utils'
import Slider from '..'

describe('Slider', () => {
  it('should show tooltip when hovering slider handler', async () => {
    const wrapper = mount(Slider, {
      propsData: {
        defaultValue: 30,
      },
      sync: false,
    })
    await asyncExpect(() => {
      wrapper.findAll('.ant-slider-handle').at(0).trigger('mouseenter')
    })
    let dropdownWrapper = null
    await asyncExpect(() => {
      dropdownWrapper = mount({
        render () {
          return wrapper.find({ name: 'Trigger' }).vm.getComponent()
        },
      }, { sync: false })
    })
    await asyncExpect(() => {
      expect(dropdownWrapper.html()).toMatchSnapshot()
      wrapper.findAll('.ant-slider-handle').at(0).trigger('mouseleave')
    })
    await asyncExpect(() => {
      dropdownWrapper = mount({
        render () {
          return wrapper.find({ name: 'Trigger' }).vm.getComponent()
        },
      }, { sync: false })
    })
    await asyncExpect(() => {
      expect(dropdownWrapper.html()).toMatchSnapshot()
    })
  })
})
