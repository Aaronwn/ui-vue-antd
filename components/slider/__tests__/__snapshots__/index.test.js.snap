// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should show tooltip when hovering slider handler 1`] = `
<div>
  <div class="ant-tooltip  ant-tooltip-placement-top" style="left: -999px; top: -1003px;">
    <div class="ant-tooltip-content">
      <div class="ant-tooltip-arrow"></div>
      <div class="ant-tooltip-inner">30</div>
    </div>
  </div>
</div>
`;

exports[`Slider should show tooltip when hovering slider handler 2`] = `
<div>
  <div class="ant-tooltip  ant-tooltip-placement-top" style="display: none; left: -999px; top: -1003px;">
    <div class="ant-tooltip-content">
      <div class="ant-tooltip-arrow"></div>
      <div class="ant-tooltip-inner">30</div>
    </div>
  </div>
</div>
`;
